waku/v2/node/peer_manager/peer_manager.nim

```diff
@@ -54,45 +64,65 @@ proc insertOrReplace(ps: PeerStorage,
 proc dialPeer(pm: PeerManager, peerId: PeerID,
               addrs: seq[MultiAddress], proto: string,
               dialTimeout = DefaultDialTimeout,
-              source = "api"
+              source = "api",
+              respectBackoff = false,
               ): Future[Option[Connection]] {.async.} =
 
   # Do not attempt to dial self
   if peerId == pm.switch.peerInfo.peerId:
     return none(Connection)
 
-  info "Dialing peer from manager", wireAddr = addrs, peerId = peerId
+  let failedAttempts = pm.peerStore[NumberFailedConnBook][peerId]
+
+  # Respect a backoff period for peers that have failed to connect
+  if respectBackoff and not pm.peerStore.canBeConnected(peerId, pm.initialBackoffInSec):
+    debug "Peer recently failed to connect, skipping dial",
+          peerId = peerid,
+          failedAttempts = failedAttempts,
+          lastFailed = pm.peerStore[LastFailedConnBook][peerId]
+    return none(Connection)
+
+  debug "Dialing peer from manager", wireAddr = addrs, peerId = peerId, failedAttempts=failedAttempts
 
   # Dial Peer
+  var deadline = sleepAsync(dialTimeout)
   let dialFut = pm.switch.dial(peerId, addrs, proto)
 
+  var reasonFailed = ""
   try:
-    # Attempt to dial remote peer
-    if (await dialFut.withTimeout(DefaultDialTimeout)):
+    await dialFut or deadline
+    if dialFut.finished():
+      if not deadline.finished():
+        deadline.cancel()
```

Comment:
Any reason not to use `dialFut.withTimeout()`?

Reply:
just reverted my changes, using again withTimeout.

i thought there was a possible race condition and wanted also to cancel the dial if the timer timed out, but noticed it didnt made much sense.